[WIP/TEST] Draw GUI controls as separate textures

ivan-mogilko opened the pull request:
This is only a test for now, so marked as wip.

The GUI controls were historically not rendered as separate objects added into the list of sprites, but instead drawn on a GUI surface. On one hand this is very simple, on another - a) engine has to redraw whole GUI when anything on it changes, b) this does not let to apply same graphical properties and effects as for other objects (accelerated transformation, shaders and so on).

This pr changes this, giving each gui control its texture and passing them through the same procedure as every other object on screen.

In terms of perfomance, in low-res games there is no much difference. In high-resolution games there's a noticable perfomance improvement (especially in unlimited fps mode), but it turned out to be not that big. There were other perfomance fixes recently, which restricted GUI redraws to only when something really changes on them, and these seem to had much more impact.
Perhaps it will be interesting to make tests with larger amount of changing controls, like multiple animated buttins in high resolution, and so on.

In any case, for now this particular change does not seem to be seriously necessary for ags3 branch (3.6.0 version). This may come handy when we want to support individual control scaling and other transformations, and/or rendering controls in the window resolution (similar to the character sprites in the corresponding mode).

---

Current issues:

1. Control contents (text, image) are clipped by the size. This is different from how the engine currently works, where text may freely exceed control frame (due to how they are drawn). Personally I think content clipping is how it is normally supposed to be, but regardless it would have to have non-clipping mode to support old behavior. This may be done by telling the "extended metrics" of the controls, precalculating the size their contents would take.
2. On contrary, the controls themselves are supposed to be clipped by their parent GUI, but with this change they are not. The solution may be to define each GUI children as a sprite batch limited to GUI-sized area when these sprites are passed into the renderer.
3. Historically GUI had 3 modes of blending controls (named "classic", "additive" and "proper"). Only "proper" would be supported by OpenGL and Direct3D without additional changes. The "classic" perhaps may be done by marking controls' sprites "opaque", but "additive" mode would require a shader. Temporarily in the first two old modes controls are drawn the old way (on GUI bitmap).

ivan-mogilko commented:
> On contrary, the controls themselves are supposed to be clipped by their parent GUI, but with this change they are not. The solution may be to define each GUI children as a sprite batch limited to GUI-sized area when these sprites are passed into the renderer.

This may be solved using a simplier solution within 3.6.0 ([example](https://github.com/adventuregamestudio/ags/pull/1393/commits/d3a2a897a2e19aef1f3f883486c603c8cd4e090d)), but will anyway require completing #967 for ags4, because guis support additional transformations there.

ericoporto commented:
Could GUIs be like viewport and camera for the GUI Controls they own? 

I don't know if it's possible since we can't ZOrder Viewports, my guess is it's not feasible with the current things in the graphics part.

ivan-mogilko commented:
> Could GUIs be like viewport and camera for the GUI Controls they own?

The difference is that viewports dont have their own graphic, but GUIs do. So there's no "texture" passed for viewport, but there is for gui. Also guis are sorted among overlays too.

> I don't know if it's possible since we can't ZOrder Viewports

Viewports have ZOrder property, but it works only among viewports atm. Sorting them among GUI would require rewriting how list of drawable items is constructed, so probably will be possible only after #967.

> having this as is (without #967), would it make merging things to ags4 harder

Merging code itself would be not difficult, but it simply won't work in ags4; the fallback method would work, but that will defeat the purpose of this feature. So if i do that, someone will have to later fully reimplement control texture rendering in ags4.

ivan-mogilko commented:
Okay, i did something similar to #967 here, now there's BeginSpriteBatch and EndSpriteBatch, and they support sub-levels. At least this works with GUI controls.

ericoporto commented:
Just a question, not something for now, just thinking an idea here, can it check after a frame ended that no begin call was left missing an end and then log a warning somewhere?

ivan-mogilko commented:
So, the rendering controls as textures currently has restrictions, which should be clear from the code:
https://github.com/adventuregamestudio/ags/pull/1393/commits/dadd6d81fb9160623a1ca8c9f4ed01b1d41a271e#diff-1d31b249a940535166d842cb1b562d67f46362d7a99e5ea8aa127f942fe24b6dR2286-R2293

These restrictions may theoretically be removed with additional changes later.
For software renderer, there's additional concern that using the old way may be faster at least in low-mid game resolutions.

ivan-mogilko commented:
So, the rendering controls as textures currently has restrictions, which should be clear from the code:
https://github.com/adventuregamestudio/ags/pull/1393/commits/667bdc939648d705fc385dedaae5e2cd80cf3b99#diff-1d31b249a940535166d842cb1b562d67f46362d7a99e5ea8aa127f942fe24b6dR2286-R2293

These restrictions may theoretically be removed with additional changes later.
For software renderer, there's additional concern that using the old way may be faster at least in low-mid game resolutions.
In regards to clipping, this is now solved for the Slider, the only control that must be drawn without clipping for now, as we decided previously. Removing this restriction is a matter of precalculating unclipped text size for labels and buttons and list boxes; that may be easy, but it's better to do some refactor and ensure this calculation is done only when the text and other properties change, and not each frame.

ivan-mogilko commented:
Got good news from @tarnos12, who has a 1920x1080 game, and this change significantly improves a scene with multiple moving gui buttons. So all this definitely was not in vain.

ivan-mogilko commented:
Rebased and squashed commits.

Added functions that calculate control content sizes for "non-clipping" mode; this allows to also draw these controls as textures.
This increases the code size, but not *too* much. On the other hand this will make the new render mode enhance running older high-resolution games too. I think this may be cut from the ags4 though, if we decide to keep gui controls strictly clipped.

ericoporto commented:
Hey, I think someone is missing in the change from the commit here: https://github.com/adventuregamestudio/ags/pull/1393/commits/0156847541c02777f6ca0b672667ea732e8845da

The error at building the Editor complains of s missing NotifyParentChanged.

ivan-mogilko commented:
In regards to inventory's white rectangle, the same bug is present in tag v.3.6.0.22.

ericoporto commented:
Hey, reading the top post I don't understand if you plan on merging this or not, but if you do, vga256 on the forums is making a game that uses lots of GUIs, so if there's any issue I am sure it will be picked quickly. So once this is merged, probably would be nice to get a new release not long after.

ivan-mogilko commented:
Updated the pr description.

ericoporto commented:
I have a minor issue in the GUIs of I Rented a Boat. Left is from this PR, right is from master (which is the intended look)
![image](https://user-images.githubusercontent.com/2244442/162839302-12cbde21-d530-47ac-b21b-08dbdda24b41.png)

ivan-mogilko commented:
I guess the "has alpha" flag may be not set correctly on the control textures.

ericoporto commented:
Damn, find another thing, still on I Rented a Boat. If you load it, in the options menu, you can click to alternate between Windowed and Fullscreen. With the current master, this works alright, but with this PR I get the following error:

![image](https://user-images.githubusercontent.com/2244442/162849234-97cd78a5-cbd8-4d6c-a441-beddad2e2618.png)

ivan-mogilko commented:
> Damn, found another thing, still on I Rented a Boat. If you load it, in the options menu, you can click to alternate between Windowed and Fullscreen. With the current master, this works alright, but with this PR I get the following error:

So, apparently this happens when buttons change visibility. I.e. similar code:
https://github.com/ericoporto/i_rented_a_boat/blob/main/I_rented_a_boat/Menu.asc#L42-L46

ericoporto commented:
when visibility changes, are any members of the `guiobjbmp` vector removed?

ericoporto commented:
Hey, I tested some more games and I could not finding anything now. I think it's safe to merge! 😃

ivan-mogilko commented:
I realized that made a mistake with the changes to sorting algorithm ([this commit in particular](https://github.com/adventuregamestudio/ags/pull/1393/commits/bbae9ce641f5b9ab26e8ca797c2f7efdf21ee0fe)); it works only while certain conditions are met. I need to rethink this and make differently.

ivan-mogilko commented:
Ok I think it's done now; although gathering overlay + gui + gui control textures has a little "unexpected" order of actions, but in general the code became more straightforward (imho). It's far from perfect, but this is trying to work within the existing system...
